### PR TITLE
[2.0] Reodering arguments for torch jit #1010 and future TF2.0 compatibility

### PR DESCRIPTION
**Pull request by thomwolf**
Torch jit (cf #1010) and TF 2.0 (cf #1104) are more strict than PyTorch on having a specific order of arguments for easy use.

This PR refactor the order of the keyword arguments to make them as natural as possible.

This will be a breaking change for people using positional order to input keyword arguments in the forward pass of the models, hence is delayed to the 2.0 release.

### CODE REVIEW

File: pytorch_transformers/modeling_bert.py [hunk 1]
```diff
@@ -957,10 +970,15 @@ def __init__(self, config):
 
         self.init_weights()
 
-    def forward(self, input_ids, token_type_ids=None, attention_mask=None, labels=None,
-                position_ids=None, head_mask=None):
-        outputs = self.bert(input_ids, position_ids=position_ids, token_type_ids=token_type_ids,
-                            attention_mask=attention_mask, head_mask=head_mask)
+    def forward(self, input_ids, attention_mask=None, token_type_ids=None,
+                position_ids=None, head_mask=None, labels=None):
```
Comment (LysandreJik) [inline pytorch_transformers/modeling_bert.py]: This seems to solve #1010

File: pytorch_transformers/modeling_gpt2.py [hunk 2]
```diff
@@ -257,14 +257,18 @@ def prune_heads(self, heads):
...
         if self.scale:
             w = w / math.sqrt(v.size(-1))
         nd, ns = w.size(-2), w.size(-1)
         b = self.bias[:, :, ns-nd:ns, :ns]
         w = w * b - 1e4 * (1 - b)
 
+        if attention_mask is not None:
+            # Apply the attention mask
+            w = w + attention_mask
+
```
Comment (LysandreJik) [inline pytorch_transformers/modeling_gpt2.py]: Nice addition!

File: pytorch_transformers/modeling_openai.py [hunk 3]
```diff
@@ -279,6 +279,10 @@ def _attn(self, q, k, v, head_mask=None):
         b = self.bias[:, :, : w.size(-2), : w.size(-1)]
         w = w * b + -1e9 * (1 - b)
 
+        if attention_mask is not None:
+            # Apply the attention mask
+            w = w + attention_mask
```
Comment (LysandreJik) [inline pytorch_transformers/modeling_openai.py]: Nice addition here too

File: pytorch_transformers/tests/modeling_bert_test.py [hunk 4]
```diff
@@ -126,8 +126,8 @@ def check_loss_output(self, result):
         def create_and_check_bert_model(self, config, input_ids, token_type_ids, input_mask, sequence_labels, token_labels, choice_labels):
             model = BertModel(config=config)
             model.eval()
-            sequence_output, pooled_output = model(input_ids, token_type_ids, input_mask)
-            sequence_output, pooled_output = model(input_ids, token_type_ids)
+            sequence_output, pooled_output = model(input_ids, attention_mask=input_mask, token_type_ids=token_type_ids)
```
Comment (LysandreJik) [inline pytorch_transformers/tests/modeling_bert_test.py]: I feel like this is clearer than before! Looks great.

**Review by LysandreJik — APPROVED** [review]
This seems to solve the problems related to JIT. I feel it's a great addition as the order of inputs now accurately depicts the order of importance.

**Comment by thomwolf** [comment]
Ok merging

**Pull request merged**